File: server/models/PG.js
import mongoose from 'mongoose';

const pgSchema = new mongoose.Schema(
  {
    title: {
      type: String,
      required: [true, 'Title is required'],
      trim: true,
      minlength: [10, 'Title must be at least 10 characters'],
      maxlength: [100, 'Title cannot exceed 100 characters'],
    },
    description: {
      type: String,
      required: [true, 'Description is required'],
      minlength: [50, 'Description must be at least 50 characters'],
    },
    price: {
      type: Number,
      required: [true, 'Price is required'],
      min: [500, 'Price must be at least ₹500'],
    },
    location: {
      city: { type: String, required: [true, 'City is required'] },
      area: { type: String, required: [true, 'Area is required'] },
      address: { type: String, required: [true, 'Address is required'] },
    },
    roomType: {
      type: String,
      enum: ['Single', 'Double', 'Triple'],
      required: [true, 'Room type is required'],
    },
    genderPreference: {
      type: String,
      enum: ['Boys', 'Girls', 'Co-ed'],
      required: [true, 'Gender preference is required'],
    },
    amenities: [
      {
        type: String,
        enum: ['WiFi', 'Food', 'AC', 'Parking', 'Laundry', 'CCTV', 'Water', 'Gym'],
      },
    ],
    images: [{ type: String }],
    contactNumber: {
      type: String,
      required: [true, 'Contact number is required'],
      match: [/^[+]?[6-9]\d{9}$/, 'Invalid Indian phone number'],
    },
    owner: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: [true, 'Owner is required'],
    },
    status: {
      type: String,
      enum: ['pending', 'approved', 'rejected', 'pending_update'],
      default: 'pending',
    },
    rejectionReason: {
      type: String,
      default: '',
    },
    isDeleted: {
      type: Boolean,
      default: false,
    },
    averageRating: {
      type: Number,
      default: 0,
      min: 0,
      max: 5,
    },
    totalReviews: {
      type: Number,
      default: 0,
    },
  },
  { timestamps: true }
);

// Indexes for performance
pgSchema.index({ status: 1, isDeleted: 1 });
pgSchema.index({ 'location.city': 1 });
pgSchema.index({ price: 1 });
pgSchema.index({ owner: 1 });

const PG = mongoose.model('PG', pgSchema);
export default PG;

File: backend/app/schemas/pg.py
from pydantic import BaseModel, Field

from app.models.enums import GenderPreference, RoomType


class PGLocation(BaseModel):
    city: str
    area: str
    address: str


class PGCreateRequest(BaseModel):
    title: str = Field(min_length=10, max_length=100)
    description: str = Field(min_length=50)
    price: float = Field(ge=500)
    city: str
    area: str
    address: str
    roomType: RoomType
    genderPreference: GenderPreference
    amenities: list[str] = []
    contactNumber: str = Field(pattern=r"^[+]?[6-9]\d{9}$")
    images: list[str] = []


class PGUpdateRequest(BaseModel):
    title: str | None = Field(default=None, min_length=10, max_length=100)
    description: str | None = Field(default=None, min_length=50)
    price: float | None = Field(default=None, ge=500)
    city: str | None = None
    area: str | None = None
    address: str | None = None
    roomType: RoomType | None = None
    genderPreference: GenderPreference | None = None
    amenities: list[str] | None = None
    contactNumber: str | None = Field(default=None, pattern=r"^[+]?[6-9]\d{9}$")
    removeImages: list[str] | None = None
    images: list[str] | None = None


class PGListQuery(BaseModel):
    city: str | None = None
    minPrice: float | None = None
    maxPrice: float | None = None
    roomType: RoomType | None = None
    gender: GenderPreference | None = None
    amenities: str | None = None
    page: int = 1
    limit: int = 9
    sort: str = "newest"